### PR TITLE
[WIP] Add distances [Euclidean, WeightedEuclidean]

### PR DESCRIPTION
**Pull request by nrbrd**
- [x] Euclidean
- [ ] WeightedEuclidean

### CODE REVIEW

**Comment by apaszke** [comment]
Is that any different from `torch.dist`? I don't think it's worth including in nn if it's already exposed as a method.

**Comment by aron-bordin** [comment]
The difference is that `EuclideanDistance` computes a vector-wise distance, where you can select the dimension to perform it.

```
>>> a = torch.rand([3,5])
>>> b = torch.rand([3,5])

>>> torch.dist(a, b)
1.5644694566726685
```

while

```
>>> F.euclidean_distance(a, b, dim=0)
0.6358
 0.8686
 0.4163
 0.7261
 0.7671
[torch.FloatTensor of size 5]
```


```
>>> F.euclidean_distance(a, b, dim=1)

 0.7050
 1.2043
 0.7073
[torch.FloatTensor of size 3]

```
I think that this should be useful for semi-supervised/unsupervised learning. But for sure, this can be easily implemented in the user space and not in torch. I implemented because I found these TODOS here: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/distance.py#L34

Now asking, should I implement the WeightedEuclidean, or it will not be necessary ?

File: torch/nn/functional.py [hunk 1]
```diff
@@ -713,6 +713,32 @@ def cosine_similarity(x1, x2, dim=1, eps=1e-8):
...
+    Shape:
+        - Input: :math:`(\ast_1, D, \ast_2)` where D is at position `dim`.
+        - Output: :math:`(\ast_1, \ast_2)` where 1 is at position `dim`.
+
+    >>> input1 = autograd.Variable(torch.randn(100, 128))
+    >>> input2 = autograd.Variable(torch.randn(100, 128))
+    >>> output = F.euclidean_distance(input1, input2)
+    >>> output.backward()
+    """
+    assert x1.size() == x2.size(), "Input sizes must be equal."
```
Comment (fmassa) [inline torch/nn/functional.py]: Should we be using `assert` here, or raise something like `ValueError`?

**Comment by fmassa** [comment]
I think `torch.dist` is very limited as pointed out by @aron-bordin , and I think we should maybe extend it like we extended `max` / `min` (long time ago) to support a dimension, and if the dimension is not passed, fall back to the old behavior. What do you think?

**Comment by apaszke** [comment]
I think adding a `dim` to `torch.dist` would be better. If `dim=0` you can already use `F.pairwise_distance`. If you feel like adding it please open a new PR. Sorry for having you wait so long 😕

**Pull request closed without merging**